glean_parser/templates/python_server.jinja2

```diff
@@ -0,0 +1,186 @@
+    ) -> None:
+        """
+        Create {{ ping|camelize }}ServerEventLogger instance.
+
+        :param str application_id: The application ID.
+        :param str app_display_version: The application display version.
+        :param str channel: The channel.
+        """
+        self._application_id = application_id
+        self._app_display_version = app_display_version
+        self._channel = channel
+
+    def _record(
+		self,
+		user_agent: str,
+		ip_address: str,
+		{% for metric_type, metrics in metrics_by_type.items() %}
+		{% if metric_type != 'event' %}
+		{% for metric in metrics %}
+		{{ metric.category }}_{{ metric.name }}: {{ metric.type|py_metric_type }},
+		{% endfor %}
+		{% endif %}
+		{% endfor %}
+		event: dict[str, Any]
+	) -> None:
+        now = datetime.now(timezone.utc)
+        timestamp = now.isoformat()
+        event["timestamp"] = int(1000.0 * now.timestamp())  # Milliseconds since epoch
+        event_payload = {
+            "metrics": {
+				{% for metric_type, metrics in metrics_by_type.items() %}
+				{% if metric_type != 'event' %}
+				"{{ metric_type }}": {
+					{% for metric in metrics %}
+					"{{ metric.category }}.{{ metric.name }}": {{ metric.category }}_{{ metric.name }},
+					{% endfor %}
+				},
+				{% endif %}
+				{% endfor %}
+			},
+            "events": [event],
+            "ping_info": {
+                # seq is required in the Glean schema, however is not useful in server context
+                "seq": 0,
+                "start_time": timestamp,
+                "end_time": timestamp,
+            },
+            # `Unknown` fields below are required in the Glean schema, however they are
+            # not useful in server context
+            "client_info": {
+                "telemetry_sdk_build": "glean_parser v{{ parser_version }}",
+                "first_run_date": "Unknown",
+                "os": "Unknown",
+                "os_version": "Unknown",
+                "architecture": "Unknown",
+                "app_build": "Unknown",
+                "app_display_version": self._app_display_version,
+                "app_channel": self._channel,
+            },
+        }
+        event_payload_serialized = json.dumps(event_payload)
+
+        # This is the message structure that Decoder expects:
+        # https://github.com/mozilla/gcp-ingestion/pull/2400
+        ping = {
+            "document_namespace": self._application_id,
+            "document_type": "{{ ping }}",
+            "document_version": "1",
+            "document_id": str(uuid4()),
+            "user_agent": user_agent,
+            "ip_address": ip_address,
+            "payload": event_payload_serialized,
+        }
+
+        ping_envelope = {
+            "Timestamp": timestamp,
+            "Logger": "glean",
+            "Type": GLEAN_EVENT_MOZLOG_TYPE,
+            "Fields": ping
+        }
+        ping_envelope_serialized = json.dumps(ping_envelope)
+
+        print(ping_envelope_serialized)
```

Comment:
Extracting this to a method sounds fine to me.
FWIW fields in the ping envelope are inspired by Mozlog which you probably don't use. We don't really need `Timestamp` and `Logger`, but other than that our infrastructure expects the log line to be a JSON with specified `Type` and data in `Fields`.

Reply:
Maybe we should just use `logging` in the template then? That would be similar to what we do in JS/Ts.

It would be something like:
```
import logging
logger = logging.getLogger(GLEAN_EVENT_MOZLOG_TYPE)
log.info(ping_envelope_serialized)
```
We'd require applications to configure the output like you do.
I guess you'd only need to configure this logger in https://github.com/mozilla/fx-private-relay/blob/5659e0ef42be097a152cba1b9153fbac2e3421d1/privaterelay/settings.py#L580-L618

Reply:
I think `print` is a decent default, requiring no configuration and you've got good test. Relay can override `emit_record` to use the log. We could address the flexibility with documentation.

Alternatively, we could add an initialization option to use logging, or try to read the logging configuration to see if there is an appropriate logger (I'd need to explore if this is possible).

Reply:
If we're just looking for the dockerflow mozlogger, which may be reasonable for Mozilla Python projects, this works:

```
def is_dockerflow_mozlog() -> bool:

    try:
        from dockerflow.logging import JsonLogFormatter
    except ImportError:
        return False
    import logging
    logger = logging.getLogger(GLEAN_EVENT_MOZLOG_TYPE)
    for handler in logger.handlers:
        if isinstance(handler.formatter, JsonLogFormatter):
            return True
    return False
```

I've pushed https://github.com/mozilla/fx-private-relay/commit/51f5e88643822b7a83e4982df0aea60bc728fb7c to show what the generated code looks like in our repo.